Leap 48in24 approaches (#1401)
exercism/elixir: exercises/practice/leap/.approaches/introduction.md

Such operation in computing is called [modulo][modulo]. 

Unlike many languages, Elixir does not have [operators][operators] for either integer division or modulo. 
Instead, it provides the [`Kernel.rem/2`][rem] and the [`Integer.mod/2`][mod] functions.

The two functions differ in how they work with negative numbers, but since, in this exercise,
all the numbers are non-negative, both could work, depending on the approach.

## General solution

To check if a year is divisible by `n`, we can do `rem(year, n) == 0`. 

Any approach to the problem will perform this check three times to see if a year is equally divisible by 4, 100 and 400.
What will differ between approaches is what Elixir features we will use to combine the checks. 

## Approach: Boolean operators

The full rules are as follows:
A year is a leap year if 
* it is divisible by 4 
* but not divisible by 100
* unless it is divisible by 400

We can use [boolean operators][boolean-operators] to combine the checks, for example, like so: 

```elixir
rem(year, 5) == 0 and not rem(year, 100) == 0 or rem(year, 400) == 0
```

> In the [boolean operators appraoch][operators-approach] we discuss the details of the solution.
> It includes variations of the operators and their precendence.

Reviewer: Some typos
```suggestion
In the [boolean operators approach][operators-approach] we discuss the details of the solution.
It includes variations of the operators and their precedence.
```

## Approach: multiple clause function

Instead of using boolean operators, we can define multiple `leap_year?/1` function clauses with different guards. 

```elixir
def leap_year?(year) when rem(year, 400) == 0, do: true
def leap_year?(year) when rem(year, 100) == 0, do: false
def leap_year?(year) when rem(year, 4) == 0, do: true
def leap_year?(_), do: false
```

> In the [multiple clause functions approach][functions-approach] we discuss why in this approach the `Integer.mod/2` function will not work. 

Reviewer: There is only one function in this exercise, `leap_year?`:
```suggestion
In the [multiple clause function approach][functions-approach] we discuss why in this approach the `Integer.mod/2` function will not work. 
```

Also, the directory name of this concept, "functions", and this link name, `[functions-approach]` (also used in other markdown files), is not appropriate. Every single approach has "functions" 😁 this name should be something distinctive about the approach. Maybe "multiple-clause-function", or shorter, "multiple-clauses"

Author: @angelikatyborska, how about `clauses` or `function-clauses` for the slug?

Reviewer: sounds good too

## Approach: control flow structures

In addition to the above two approaches, control flow structures offer a number of solutions.
Here are two examples using `if` and `case`. 

```elixir
if rem(year, 100) == 0 do
    rem(year, 400) == 0
else 
    rem(year, 4) == 0
end
```

```elixir
case { rem(year, 400), rem(year, 100), rem(year, 4) } do
    { 0, _, _ } -> true
    { _, 0, _ } -> false
    { _, _, 0 } -> true

>     { _, _, _ } -> false

Reviewer: Like @Tuxified mentioned, the most idiomatic way should be presented here:
```suggestion
    true -> false
```

end
```

We discuss these and other solutions depending on various control flow structures in the [control flow structures approach][flow-approach].

[modulo]: https://en.wikipedia.org/wiki/Modulo
[operators]: https://hexdocs.pm/elixir/operators.html
[rem]: https://hexdocs.pm/elixir/Kernel.html#rem/2
[mod]: https://hexdocs.pm/elixir/Integer.html#mod/2
[boolean-operators]: https://hexdocs.pm/elixir/operators.html#general-operators
[operators-approach]: https://exercism.org/tracks/elixir/exercises/leap/approaches/operators
[clause-approach]: https://exercism.org/tracks/elixir/exercises/leap/approaches/clauses

> [flow-approach]: https://exercism.org/tracks/elixir/exercises/leap/approaches/cond

Reviewer: This one broken link snuck in to `main`. Could you open a PR with a fix? If you do it, I can just merge, but if I do it, I will need to wait for another person with write rights to review because I can't approve my own PRs 😇

# Introduction

Every fourth year is a leap year (with some exceptions), but let's consider this one condition first. 

To solve the Leap problem, we must determine if a year is evenly divisible by a number or if a reminder of an integer division is zero. 
Such operation in computing is called [modulo][modulo]. 

Unlike many languages, Elixir does not have [operators][operators] for either integer division or modulo. 
Instead, it provides [`rem/2`][rem] guard and the [`Integer.mod/2`][mod] function.

The two functions differ in how they work with negative numbers, but since, in this exercise,

> all the numbers are non-negative, both could work, depending on the approach.

Author: I used the "non-negative" here very intentionally. Positives start at 1, non-negative start at 0, so the zero is included. Probably a better name for the `n` parameter and for the description would be to say 'divisor' as this is technically what it is, and where the behaviour of the two functions differ.

> all the numbers are non-negative, both could work, depending on the approach.

Reviewer: alright, no problem, we can keep it

> ## General Solution

Reviewer: Can you update all headings to only have the first word capitalized instead of all? That the style of headings we use in this repo
```suggestion
## General solution
```

To check if a year is divisible by `n`, we can do `rem(year, n) == 0`. We can define a function to make the intent clearer. 

```elixir
defp divides?(number, divisor), do: rem(number, divisor) == 0 

> ```

Author: I was inspired by my solution to the problem I submitted in the Clojure track. I agree that many of the example or community solutions don't represent anything close to ideal. Sometimes, I see things starred the most, which I would consider an anti-pattern rather than a solution worth showing. 

I think that the approaches are a perfect place where the solutions can be curated to things worth showing rather than cataloguing what's possible. 

As to the divdes? The gain is not big for experienced developers, but for those less experienced, I think it might help with showing intent. When it comes to naming things I'm really inspired by [Kevlin Henny's Seven Ineffective Coding Habits of Many Programmers talk](https://www.youtube.com/watch?v=SUIUZ09mnwM). For me, the code is primarily an intent communication tool. 

But I'll think a bit more about it as I write out the details of the approaches.

Any approach to the problem will perform this check three times to check if a year is equally divisible by 4, 100 and 400.
What will differ is what Elixir features we will use to combine the checks. 

## Approach: Boolean Operators

The full rules are as follows:
A year is a leap year if 
* it is divisible by 4 
* but not divisible by 100
* unless it is divisible by 400

We can use [boolean operators][boolean-operators] to combine the checks, for example, like so: 

```elixir
divides?(year, 400) or (not(divides?(year, 100))) and divides?(year, 4)
```
In the [boolean operators appraoch][operators-approach] we discuss the details of the solution.
It includes variations of the operators and their precendence.

> ## Approach: Function Guards

Reviewer: This approach should be called 
```suggestion
## Approach: Multiple clause function
```

Guards can be used either in a `case` expression or in multiple clause functions.

Instead of using boolean operators, we can define multiple `leap_year?/1` functions with different guards. 

> We can use the order of the definitions to ensure correct check.

Reviewer: This sounds wrong somehow, but I can't quite define it... "we can (...)" sounds like it's optional? But it's not optional to put the clauses in the right order, the order always matters. Maybe we can remove this sentence from here
```suggestion
```

and then in the detailed approach description mention something like:
````
Remember that the order of function clauses matters, and only the first clause that matched a specific function call will get executed. All the other clauses will be ignored. This means that we **don't** need to repeat out checks:
```elixir
# Don't copy this!
def leap_year?(year) when rem(year, 400) == 0, do: true
def leap_year?(year) when rem(year, 100) == 0 and not rem(year, 400) == 0, do: false
def leap_year?(year) when rem(year, 4) == 0 and not rem(year, 100) == 0 and not rem(year, 400) == 0, do: true
def leap_year?(year) when not rem(year, 4) == 0 and not rem(year, 100) == 0 and not rem(year, 400) == 0,, do: false
```
````

```elixir
def leap_year?(year) when rem(year, 400) == 0, do: true
def leap_year?(year) when rem(year, 100) == 0, do: false
def leap_year?(year) when rem(year, 4) == 0, do: true
def leap_year?(_), do: false
```

In the [functions with guards approach][guards-approach] we discuss why in this approach the `Integer.mod/2` function will not work. 

## Approach: Using cond

> Similarly to the functions with guards, the order of the checks can be done inside a function with `cond` expression. 

Reviewer: "the order of the checks can be done..." sounds wrong somehow. What about:
```suggestion
Similarly to the multiple clause function approach, we can also use a `cond` expression. 
```

```elixir
cond do
    divides?(year, 400) -> true
    divides?(year, 100) -> false
    divides?(year, 4) -> true
    true -> false
end
```

We discuss this briefly in the [cond approach][cond-approach]

> ## Approach: Using case

Reviewer: This approach should be called
```suggestion
## Approach: Using case on a tuple
```

There are different solutions that could be called "using case", for example any `cond` can be rewritten to a `case` (but this one is not very good, so let's not describe it 😉)

```elixir
case year do
  year when rem(year, 400) == 0 -> true
  year when rem(year, 100) == 0 -> false
  year when rem(year, 4) == 0 -> true
  _ -> false
end
```

Author: I plan to show multiple options, similar to what I've done in the boolean operators approach. the `on a tuple` will be just one variant.

Reviewer: Hmm... I'm not entirely sure what the philosophy of writing Exercism approaches is supposed to be because I don't really read them. Are they supposed to cover every possible way of solving an exercise, or just the most common or interesting ones?

Because if we only want the common ones, then "case on a tuple" is the only case-based approach that I would suggest showing, based on what the community solutions tend to be. The rest of the case approaches that I see in the current version of the `exercises/practice/leap/.approaches/flow/content.md` file are less appropriate uses of `case` that should have been an `if` or a `cond` instead (like you point out already).
